src/keygen/keyshare.rs

```diff
@@ -289,4 +289,71 @@ mod tests {
         let bytes = [KEYSHARE_TAG, &0usize.to_le_bytes()].concat();
         assert!(KeySharePrivate::try_from_bytes(bytes).is_err());
     }
+    use gmp_mpfr_sys::gmp;
+    use libpaillier::unknown_order;
+    use rand::{rngs::StdRng, SeedableRng};
+    use std::{
+        mem::{align_of, size_of},
+        slice,
+    };
+
+    use crate::gmp_zeroize::setup_zeroize;
+
+    #[allow(warnings)]
+    #[test]
+    fn zeroize_works() {
+        // Set up automatic zeroization of GMP memory.
+        setup_zeroize();
+
+        // Generate a secret.
+        let rng = &mut StdRng::from_seed([0; 32]);
+        let mut share = KeySharePrivate::random(rng);
+
+        // Pre-hack validation.
+        assert!(
+            size_of::<unknown_order::BigNumber>() == size_of::<gmp::mpz_t>() &&
+            align_of::<unknown_order::BigNumber>() == align_of::<gmp::mpz_t>(),
+            "unknown_order::BigNumber and rug::Integer should be a transparent wrapper of gmp::mpz_t."
+        );
+
+        // Dig down the stack of wrappers to get to the actual storage.
+        let limb_size = size_of::<gmp::limb_t>();
+        let data = unsafe {
+            let mpz = &*(&share.x as *const unknown_order::BigNumber as *const gmp::mpz_t);
```

Comment:
Question about this line: You are using raw pointer casting to cast our `KeySharePrivate` to a `mpz_t`. And then accessing the `d` field. I'm surprised this isn't undefined behavior? Could you elaborate on why this operation is okay?

Reply:
This is indeed UB in this test because `unknown_order::Bn` is not `#[repr(transparent)]`. Hence the before/after validation to catch possible future changes in dependencies or compiler. In that case one will need to find another way to test this.

The alternatives I know of are 1) to not test at all, or 2) to fork the dependency to expose a `pub fn inner()` method.

Reply:
Makes sense. We should leave the test in. Thanks for confirming the possibility of UB. Good to know and keep an eye out for!